src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,268 @@
+#include "AutomaticScheduling.h"
+#include "Bounds.h"
+#include "Expr.h"
+#include "FindCalls.h"
+#include "Func.h"
+#include "Schedule.h"
+#include "Simplify.h"
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+namespace Halide {
+namespace Internal {
+
+using std::set;
+using std::string;
+using std::vector;
+using std::map;
+
+namespace {
+
+const unsigned UNDEFINED_FOOTPRINT_SIZE = ~0U;
+
+/** Convenience class representing the callgraph for a pipeline. */
+class CallGraph {
+public:
+    CallGraph(Function root) {
+        set<string> visited;
+        construct(root, visited);
+    }
+
+    /** Return list of functions directly calling function f. */
+    const vector<Function> &callers(Function f) {
+        map<string, vector<Function> >::const_iterator I = call_to_caller.find(f.name());
+        if (I == call_to_caller.end()) {
```

Comment:
I think this logic would be equivalent to simply:

    return call_to_caller[f.name()];

std::map's operator[] would simply return a default constructed vector<Function> if the key doesn't exist.